locust/runners.py

```diff
@@ -40,6 +41,10 @@ def __init__(self, locust_classes, options):
         self.state = STATE_INIT
         self.hatching_greenlet = None
         self.stepload_greenlet = None
+        self.current_cpu_usage = 0
+        self.cpu_threshold_exceeded = False
+        self.slave_cpu_threshold_exceeded = False
+        gevent.spawn(self.monitor_cpu)
```

Comment:
Hmm, since we don't store any reference to this greenlet, it won't get killed. We should probably change so that the `LocalLocustRunner.greenlet` is a `gevent.pool.Group` instance (just like `MasterLocustRunner` and `SlaveLocustRunner`) and then make sure the CPU monitor greenlet is spawned from this group.

Reply:
Hmm... I'm not sure how to do that. LocustRunner is supposed to be a singleton, so it shouldnt really matter, right? (not that we want to be sloppy :)

If you think this is important, would you mind taking a look yourself?

Reply:
At the moment, it's a singleton when started normally (through main.py). But we do create multiple runner instances within the tests.

Also, it's been proposed, and I think it's a good idea, to work towards an API where one can run Locust programatically, in which case I think the design will be much cleaner if we group the spawned greenlets together (that one can join) and make sure they are killed together.

I can definitely take a look!

Reply:
Thanks!

Reply:
I've now changed so that we spawn all greenlets through the runner instance's `greenlet` attribute which is a `gevent.pool.Group` instance.

I also added a test for the CPU warning (and changed so that we use a constant for the montoring interval, so that I could decrease the run time of the test).